packages/firestore/src/local/local_store.ts

```diff
@@ -783,51 +800,84 @@ export class LocalStore {
-      return this.queryCache
-        .getQueryData(txn, query)
-        .next((queryData: QueryData | null) => {
-          assert(
-            queryData != null,
-            'Tried to release nonexistent query: ' + query
-          );
-          const targetId = queryData!.targetId;
-          const cachedQueryData = this.queryDataByTarget[targetId];
-
-          // References for documents sent via Watch are automatically removed when we delete a
-          // query's target data from the reference delegate. Since this does not remove references
-          // for locally mutated documents, we have to remove the target associations for these
-          // documents manually.
-          const removed = this.localViewReferences.removeReferencesForId(
-            targetId
+      return this.getQueryData(txn, query).next(queryData => {
+        assert(
+          queryData != null,
+          'Tried to release nonexistent query: ' + query
+        );
+        const targetId = queryData!.targetId;
+
+        // References for documents sent via Watch are automatically removed
+        // when we delete a query's target data from the reference delegate.
+        // Since this does not remove references for locally mutated documents,
+        // we have to remove the target associations for these documents
+        // manually.
+        const removed = this.localViewReferences.removeReferencesForId(
+          targetId
+        );
+        delete this.queryDataByTarget[targetId];
+        if (!keepPersistedQueryData) {
+          return PersistencePromise.forEach(removed, (key: DocumentKey) =>
+            this.persistence.referenceDelegate.removeReference(txn, key)
+          ).next(() =>
+            this.persistence.referenceDelegate.removeTarget(txn, queryData!)
           );
-          delete this.queryDataByTarget[targetId];
-          if (!keepPersistedQueryData) {
-            return PersistencePromise.forEach(removed, (key: DocumentKey) =>
-              this.persistence.referenceDelegate.removeReference(txn, key)
-            ).next(() =>
-              this.persistence.referenceDelegate.removeTarget(
-                txn,
-                cachedQueryData
-              )
-            );
-          } else {
-            return PersistencePromise.resolve();
-          }
-        });
+        } else {
+          return PersistencePromise.resolve();
+        }
+      });
     });
   }
 
   /**
    * Runs the specified query against all the documents in the local store and
    * returns the results.
    */
-  executeQuery(query: Query): Promise<DocumentMap> {
+  executeQuery(query: Query): Promise<DocumentMap>;
+  /**
+   * Runs the specified query against the local store and returns the results,
+   * potentially taking advantage of the provided query data and the set of
+   * remote document keys.
+   */
+  executeQuery(
+    query: Query,
+    queryData: QueryData | null,
+    remoteKeys: DocumentKeySet
```

Comment:
Sorry, but I'm not really a fan:
1. I think the default behavior should be a full collection scan and you should have to explicitly opt into the code paths that take advantage of optimizations that might not always be safe.
2. "needsRefill" is specifically used for a different purpose in the SDK and so we shouldn't overload it...  and I don't think it's a good term for this situation anyway (what are we "refilling"?).
3. I still don't like passing in all of `QueryData`.  It's kind of a grab bag data structure that was born out of convenience (it's hard to describe to somebody what it represents).  Propagating it to more layers of our API does not seem beneficial (there's probably some OO principle about giving methods the minimum data they need to accomplish their task or something...?)

I'm not 100% sure what it means for this to be implementation agnostic (or else why it's important).  I think if you wrapped this up in a `CachedResultsInfo` or whatever it would become very clear why e.g. `SimpleQueryEngine` just ignores it... whereas right now it's not clear what a QueryEngine implementer should / shouldn't do with remoteKeys and queryData.

Reply:
FWIW (you're still free to push back / suggest alternatives) - Fleshing out my proposal a bit more, I think I'd maybe change CachedResultsInfo.lookup(targetId) to CachedResultsInfo.unknown(targetId) and add CachedResultsInfo.None, which would be the default behavior if no CachedResultsInfo was passed.

Reply:
I updated this based on our offline discussion:

- `executeQuery()` now returns the remoteKeys, which means that there are now only two cases this method needs to address (Index-Free: yes or no). We don't need to pass remote keys since the method returns them. As discussed, this adds slight overhead to limit Query refills.

- This change also solves a potential multi-tab race since the set of remote keys and the set of remote documents are now fetched in a single transaction.

- I also changed the QueryEngine to only take the lastLimbFreeSnapshotVersion instead of QueryData.